core/shared/src/main/scala/fs2/concurrent/TimedPull.scala

```diff
@@ -0,0 +1,289 @@
+              .map(_.map(_.map(toTimedPull)))
+
+          def startTimer(t: FiniteDuration): Pull[F, INothing, Unit] = Pull.eval {
+            newTimeout(t).flatMap(t => time.set(t.some))
+          }
+        }
+
+        pull(toTimedPull(output)).stream
+      }
+    }
+  }
+
+  def groupWithin[O](s: Stream[IO, O], n: Int, t: FiniteDuration) =
+    TimedPull.go[IO, O, Chunk[O]] { tp =>
+      def emitNonEmpty(c: Chunk.Queue[O]): Pull[IO, Chunk[O], Unit] =
+        if (c.size > 0) Pull.output1(c.toChunk)
+        else Pull.done
+
+      def resize(c: Chunk[O], s: Pull[IO, Chunk[O], Unit]): (Pull[IO, Chunk[O], Unit], Chunk[O]) =
+        if (c.size < n) s -> c
+        else {
+          val (unit, rest) = c.splitAt(n)
+          // TODO use Chunq.queue here instead of aggregating via emission?
+          resize(rest, s >> Pull.output1(unit))
+        }
+
+      def go(acc: Chunk.Queue[O], tp: TimedPull[IO, O]): Pull[IO, Chunk[O], Unit] =
+        tp.uncons.flatMap {
+          case None => emitNonEmpty(acc)
+          case Some((e, next)) =>
+            e match {
+              case Left(_) =>
+                emitNonEmpty(acc) >> tp.startTimer(t) >> go(Chunk.Queue.empty, next)
+              case Right(c) =>
+                val newAcc = acc :+ c
+                if (newAcc.size < n)
+                  go(newAcc, next)
+                else {
+                  val (toEmit, rest) = resize(newAcc.toChunk, Pull.done)
+                  toEmit >> tp.startTimer(t) >> go(Chunk.Queue(rest), next)
+                }
+            }
+        }
+
+      tp.startTimer(t) >> go(Chunk.Queue.empty, tp)
+    }.apply(s)
+
+  import cats.effect.unsafe.implicits.global
+
+  def ex = {
+    def s(as: Int*): Stream[IO, Int] = Stream(as:_*)
+    def t(d: FiniteDuration) = Stream.sleep_[IO](d)
+
+    s(1,2,3,4,5) ++ s(6,7,8,9,10) ++ t(300.millis) ++ s(11,12,13) ++ t(200.millis) ++ s(14) ++ s(5,5,5) ++ t(600.millis) ++ s(15, 16, 17) ++ s(18, 19, 20, 21, 22) ++ t(1.second)
+  }
+
+  def t = ex.groupWithin(5, 1.second).debug().compile.toVector.unsafeRunSync()
+  def tt = groupWithin(ex, 5, 1.second).debug().compile.toVector.unsafeRunSync()
+
+
+
+  // components:
+  //  a queue of chunks and timeouts
+  //  a mechanism for resettable timeouts
+  //  a way to avoid reacting to stale timeouts
+
+  // problematic interleaving
+  //  go is iterating and accumulating
+  //  the stream enqueues a chunk
+  //  the timeout enqueues
+  //  go pulls a chunk
+  //  go reaches limit, resets and starts another timeout, then recurs
+  //  go receives stale timeout, needs to be able to discard it
+  //  the state tracking the current timeout needs to be set before the next iteration of go
+
+  // if we express timedUncons and startTimer as `F`, that guarantee is hard to achieve
+  // cause they could be concurrent. We could express both as Pulls, forcing a sync loop.
+  // need to add a note that the `unconsTimed` `Pull`, unlike `uncons`, is not idempotent
```

Comment:
Can you elaborate at this? For me it's not obvious what the cause or implication of this is

Reply:
sure, I need some time to formulate a response, it's bit subtle (which is why I noted it down first thing).
As a start, try and get familiar with the stale timeout check in the old `groupWithin`

Reply:
ah wait, do you mean the idempotency comment? That's not true anymore, it is idempotent now, since it's expressed in terms of `uncons1`. I thought you meant why using `Pull` in TimedPull rather than `F[Unit]` and `F[Option[Either[Timeout, Chunk[A]]]]`, that's the subtle bit, but it's settled now

Reply:
I was referring to the idempotency comment, yes

Reply:
Right, so I knew I wanted to have `Pull` to forbid concurrency in the api (that's the thing I thought you were asking), and at first I was thinking about having an explicit queue and do `uncons = Pull.eval(dequeue1)`, which would mean however that `uncons >> uncons` would get different element, whereas the normal Pull.uncons is idempotent. I thought it didn't matter too much anyway since it's generally used linearly, but I was happier when I found out I could avoid the explicit queue altogether and rely on `merge` only, instead.